bullet/src/Base.hh

```diff
@@ -13,251 +13,294 @@
+  btRigidBody *childBody;
+  btRigidBody *parentBody;
+  ::sdf::JointType type;
   ignition::math::Vector3d axis;
+  std::unique_ptr<btTypedConstraint> constraint;
 };
 
-inline btMatrix3x3 convertMat(Eigen::Matrix3d mat)
-{
-  return btMatrix3x3(mat(0, 0), mat(0, 1), mat(0, 2),
-                     mat(1, 0), mat(1, 1), mat(1, 2),
-                     mat(2, 0), mat(2, 1), mat(2, 2));
-}
+/////////////////////////////////////////////////
+class Base : public Implements3d<FeatureList<Feature>> {
+  using EntityPtr =
+      std::variant<World *, Model *, Link *, Joint *, Collision *>;
 
-inline btVector3 convertVec(Eigen::Vector3d vec)
-{
-  return btVector3(vec(0), vec(1), vec(2));
-}
+  /// \brief Only one world allowed
+  public: std::unique_ptr<World> world;
 
-inline Eigen::Matrix3d convert(btMatrix3x3 mat)
-{
-  Eigen::Matrix3d val;
-  val << mat[0][0], mat[0][1], mat[0][2],
-         mat[1][0], mat[1][1], mat[1][2],
-         mat[2][0], mat[2][1], mat[2][2];
-  return val;
-}
+  /// \ brief The World ID
+  public: std::size_t worldId;
 
-inline Eigen::Vector3d convert(btVector3 vec)
-{
-  Eigen::Vector3d val;
-  val << vec[0], vec[1], vec[2];
-  return val;
-}
+  /// \brief Hashmap for fast access to an Element given the ID
+  public: std::unordered_map<std::size_t, EntityPtr> entities;
 
-class Base : public Implements3d<FeatureList<Feature>>
-{
-  public: std::size_t entityCount = 0;
+  /// \brief Hashmap for fast access to a Container given its child ID
+  public: std::unordered_map<std::size_t, EntityPtr> containers;
 
-  public: inline std::size_t GetNextEntity()
-  {
-    return entityCount++;
-  }
+  /// \brief Counter for generating IDs
+  private: std::size_t entityCount = 0;
+
+  /// \brief Gets the next available ID
+  public: inline std::size_t GetNextEntity() { return entityCount++; }
 
-  public: inline Identity InitiateEngine(std::size_t /*_engineID*/) override
-  {
+  /// \brief Add a World Entity to this Base object.
+  public: inline Identity InitiateEngine(std::size_t /*_engineID*/) override {
     const auto id = this->GetNextEntity();
     assert(id == 0);
-
     return this->GenerateIdentity(0);
   }
 
-  public: inline std::size_t idToIndexInContainer(std::size_t _id) const
-  {
-    auto it = this->childIdToParentId.find(_id);
-    if (it != this->childIdToParentId.end())
-    {
-      std::size_t index = 0;
-      for (const auto &pair : this->childIdToParentId)
-      {
-        if (pair.first == _id && pair.second == it->second)
-        {
-          return index;
-        }
-        else if (pair.second == it->second)
-        {
-          ++index;
-        }
-      }
-    }
-    // return invalid index if not found in id map
-    return -1;
+  /// \brief Add a World Entity to this Base object.
+  public: inline Identity AddWorld(std::unique_ptr<World> &&_world) {
+    auto id = this->GetNextEntity();
+    this->world = std::move(_world);
+    auto identity = this->GenerateIdentity(id, nullptr);
```

Comment:
Note how I'm using `nullptr` for the generated Identity. The function `GenerateIdentity` only takes a `shared_ptr` which means that it will also manage the lifetime of the related object. Since we don't want this, we could potentially add a version of `GenerateIdentity` that takes a raw pointer instead?